platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,762 @@
+                // to make sure the descriptor has a long enough lifetime?
+                unsafe { self.prepare_dma(&descriptors[0], *cnt as u32) };
+            }
+            _ => (),
+        }
+
+        // Set the argument to be sent
+        self.smhc
+            .smhc_cmdarg
+            .write(|w| unsafe { w.bits(params.argument) });
+
+        // Configure and start command
+        self.smhc.smhc_cmd.write(|w| {
+            w.cmd_load().set_bit();
+            w.wait_pre_over().wait();
+            w.stop_cmd_flag().bit(auto_stop);
+            w.data_trans().bit(data_trans);
+            w.trans_dir().bit(trans_dir);
+            w.chk_resp_crc().bit(chk_resp_crc);
+            w.long_resp().bit(long_resp);
+            w.resp_rcv().bit(resp_rcv);
+            w.cmd_idx().variant(cmd_idx);
+            w
+        });
+
+        guard.data.state = State::WaitForCommand;
+        // Now wait for completion or error interrupt
+        guard.wait_for_irq().await;
+        tracing::trace!("SMHC operation completed");
+        let res = if let Some(error) = guard.data.err.take() {
+            tracing::warn!(?error, "SMHC error");
+            Err(sdmmc::Error::Other) // TODO
+        } else {
+            if long_resp {
+                Ok(sdmmc::Response::Long([
+                    self.smhc.smhc_resp0.read().bits(),
+                    self.smhc.smhc_resp1.read().bits(),
+                    self.smhc.smhc_resp2.read().bits(),
+                    self.smhc.smhc_resp3.read().bits(),
+                ]))
+            } else {
+                Ok(sdmmc::Response::Short {
+                    value: self.smhc.smhc_resp0.read().bits(),
+                    data: match core::mem::replace(&mut guard.data.op, SmhcOp::None) {
+                        SmhcOp::Read { buf, .. } => Some(buf),
+                        _ => None,
+                    },
+                })
+            }
+        };
+        res
+    }
+}
+
+impl IsrData {
+    #[must_use]
+    fn lock<'a>(&'a self, smhc: &'a smhc::RegisterBlock) -> SmhcDataGuard<'a> {
+        // disable interrupts while holding the guard.
+        smhc.smhc_ctrl.modify(|_, w| w.ine_enb().disable());
+        let data = unsafe { &mut *(self.data.get()) };
+        SmhcDataGuard { data, smhc }
+    }
+}
+
+impl Drop for SmhcDataGuard<'_> {
+    fn drop(&mut self) {
+        self.smhc.smhc_ctrl.modify(|_, w| w.ine_enb().enable());
+    }
+}
+
+impl SmhcDataGuard<'_> {
+    async fn wait_for_irq(&mut self) {
+        let mut waiting = false;
+        futures::future::poll_fn(|cx| {
+            if waiting {
+                self.smhc.smhc_ctrl.modify(|_, w| w.ine_enb().disable());
+                return Poll::Ready(());
+            }
+
+            self.data.waker = Some(cx.waker().clone());
+            waiting = true;
+            self.smhc.smhc_ctrl.modify(|_, w| w.ine_enb().enable());
+
+            Poll::Pending
+        })
+        .await;
+    }
+}
```

Comment:
not something we need to address in this PR, but i kinda wonder if there's a nice way to factor out this pattern of an "IRQ lock", since it now shows up in two different driver impls...we could maybe add a trait for locking/unlocking an IRQ and implement the lock stuff on top of that?

Reply:
Yeah, I think this pattern might show up a lot!